### ISSUE TITLE
Mapping vertex buffer fails after using SetMorphWeight

### ISSUE THREAD

**Issue by henningpohl** [urho3d/Urho3D#2495]
I'm trying to use blend shapes, but it's not working for me. The model doesn't update. 

I stepped through the code and when I change the weights with `SetMorphWeight` that sets `morphsDirty_` and subsequently in `UpdateGeometry` the `UpdateMorphs` method is called. So far so good. When I continue through that it also seems to work. However an error then occurs inside the call to `buffer->Unlock()`. More specifically, that defers to `SetDataRange()`, where a `MapBuffer()` call then fails.

I'm using Direct3d11 and the corresponding implementation is:
`HRESULT hr = graphics_->GetImpl()->GetDeviceContext()->Map((ID3D11Buffer*)object_.ptr_, 0, discard ? D3D11_MAP_WRITE_DISCARD :
            D3D11_MAP_WRITE, 0, &mappedData);`
This results in an HRESULT 0x80070057 error -> E_INVALIDARG. I am not sure what is at fault here. The ID3D11Buffer pointer seems to be ok. At least it's not null.

**Comment by Modanung** [comment]
@henningpohl How did you create the model that contains the shape keys/morphs; did you use AssImp, the Blender exporter or something different?

**Comment by henningpohl** [comment]
I used the Blender exporter. The model does renders fine before I call `SetMorphWeight`, so at least the base mesh is not broken in any way. I also checked the mtl file (per https://urho3d.github.io/documentation/HEAD/_file_formats.html) and everything seems to be in place.

**Comment by Modanung** [comment]
Have you tried reproducing the issue under OpenGL?

**Comment by henningpohl** [comment]
No, I have not. I first encountered this issue not with Urho3d, but with UrhoSharp and the C++ version already is an attempt to replicate this behavior. But I'll try to check tomorrow whether I can get it working with OpenGL.

**Comment by Modanung** [comment]
I'm assuming you are building the latest master, and not using the precompiled library.
Can you affirm this assumption?

**Comment by henningpohl** [comment]
Yes, I compiled Urho3D myself (with Visual Studio 2017)

**Comment by henningpohl** [comment]
Just checked and yes, when compiled for OpenGL it works.

**Comment by henningpohl** [comment]
Any idea what to look for to make it work in Direct3D? For the project I'm working on OpenGL is not an option.

**Comment by henningpohl** [comment]
I also just tried with Direct3D9 and there it also fails, but in a different way. After calling `SetMorphWeight`, the model disappears and I get continuous "Failed to create vertex declaration" (0x80004005) errors.

**Comment by Modanung** [comment]
> Any idea what to look for to make it work in Direct3D?

I do not - for me D3D is not an option - but I'm sure the information you provided will be helpful to anyone that _can_ help you further and may look into this.

**Comment by SuperWangKai** [comment]
I wonder if it is because the VB is created without "dynamic"

Make sure the VB's `dynamic_` is `true`.
```c++
/// Dynamic flag.
bool dynamic_{};
```
